# Validación en entorno industrial

````{admonition} Resumen 
:class: tip

Este informe presenta la validación en entorno productivo real del sistema FLATCLASS, el prototipo automático de clasificación de alevines de lenguado (*Solea senegalensis*) basado en visión artificial. La validación se realizó en las instalaciones de Satistela S.A. ([Grupo Sea8](https://https://seaeight.eu/contacto/)) en Póvoa de Varzim (Portugal), mediante un protocolo multi-experimento diseñado para cuantificar: 1) la precisión en la clasificación por intervalos de talla (pequeño, mediano y grande), y 2) la exactitud en la estimación individual de biomasa de cada individuo detectado. El estudio se apoya en 1.250 observaciones (5 experimentos, 250 peces) y evalúa: exactitud, precisión, estabilidad inter-experimento, acuerdo metrológico (Bland–Altman), heterocedasticidad y un análisis Monte Carlo del riesgo de error de clasificación inducido por la incertidumbre dimensional.

**Entregable**: E6.1  
**Versión**: 1.0  
**Autor**: Javier Álvarez Osuna  
**Email**: javier.osuna@fishfarmfeeder.com  
**ORCID**: [0000-0001-7063-1279](https://orcid.org/0000-0001-7063-1279)  
**Licencia**: CC-BY-4.0  
**Código proyecto**: IG408M.2025.000.000072

```{figure} .././assets/FLATCLASS_logo_publicidad.png
:width: 100%
:align: center
```
````

## Objetivo

El **objetivo general** planteado en la validación industrial de FLATCLASS reside en comprobar hasta qué punto el sistema de visión artificial desarrollado en el {doc}`PT-1 <content/01/Modulo-3>` es capaz de clasificar de forma fiable los juveniles de lenguado en las tres diferentes tallas (pequeño, normal, grande), teniendo en cuenta la incertidumbre inherente a las mediciones realizadas en condiciones reales de operación.

Para ello, se han planteado los siguientes **objetivos específicos** orientados a analizar, de manera estructurada y cuantitativa, los distintos factores que condicionan la fiabilidad del proceso de clasificación. Estos objetivos permiten descomponer el problema global en aspectos medibles relacionados con la calidad metrológica de las estimaciones dimensionales, su estabilidad temporal, el grado de acuerdo con las medidas de referencia y el impacto práctico de los errores de medición sobre la decisión final de asignación de talla.

 1. Caracterizar estadísticamente la exactitud y precisión de las mediciones dimensionales (longitud y anchura) obtenidas mediante el sistema de visión artificial, comparándolas con las medidas de referencia realizadas en laboratorio, con el fin de cuantificar el error medio, la dispersión y la presencia de posibles sesgos sistemáticos.
 2.	Evaluar el grado de acuerdo metrológico entre las medidas estimadas y las reales, utilizando análisis de Bland–Altman y métricas complementarias, para interpretar la incertidumbre del sistema directamente en unidades físicas relevantes para la clasificación por talla.
 3.	Analizar la estabilidad temporal del sistema de medición y clasificación, comparando el comportamiento estadístico de los errores a lo largo de varios experimentos consecutivos realizados en días distintos, con el objetivo de detectar posibles derivas o variaciones operativas.
 4.	Examinar la relación entre el tamaño del individuo y la magnitud del error de medida, identificando posibles efectos de heterocedasticidad y dependencias con la talla que puedan influir en la fiabilidad de la clasificación en distintos rangos de tamaño.
 5.	Evaluar el desempeño del peso inferido como variable secundaria, analizando su precisión y variabilidad, y determinar su idoneidad para la estimación de biomasa agregada por clase de talla, en lugar de su uso como criterio primario de clasificación individual.
 6.	Cuantificar el impacto práctico de la incertidumbre dimensional sobre la decisión de clasificación por talla, mediante simulaciones Monte Carlo que permitan estimar la probabilidad de cambio de clase inducida por los errores de medición.
 7.	Identificar los rangos de talla más sensibles a la inestabilidad de clasificación, especialmente en torno a los umbrales entre clases, con el fin de caracterizar las zonas de mayor riesgo de asignación incorrecta.
 8.	Proponer criterios estadísticos para el diseño y validación de los umbrales de clasificación por talla, basados en percentiles del error dimensional, que permitan reducir el riesgo de inestabilidad de clasificación en aplicaciones operativas.
 9.	Proporcionar una base metodológica reproducible para la validación de sistemas de clasificación por visión artificial, integrando análisis metrológico, estadístico y de toma de decisiones, aplicable a otros contextos industriales y especies acuícolas.

## Metodología

### Diseño experimental

El dataset contiene **1.250 registros** correspondientes a **cinco experimentos realizados en días consecutivos** (`exp = 1..5`), con **250 juveniles por experimento**. Cada individuo se mide en cinta mediante un sistema 
de visión artificial que entrega dimensiones de **longitud** y **anchura** en píxeles que mediante conversión calibrada se registran en milímetros (`L_av`, `A_av`). A partir de estas dimensiones se produce la inferencia de peso en tiempo real (`peso_pred`)mediante el modelo alométrico previamente validado (PT-3). Como referencia, se registra **longitud real** (`L_real`), **anchura real** (`A_real`) mediante tallado manual en medición única y **peso real** (`peso_g`) mediante báscula.

Las condiciones experimentales (cámara, iluminación, calibración geométrica, procedimiento) fueron idénticas en los cinco días, permitiendo evaluar la estabilidad temporal del sistema.

```{admonition} Objetivo del sistema
:class: important
El objetivo operativo principal es la **clasificación por talla** en tres categorías discretas (pequeño, mediano, grande) basada principalmente en las **dimensiones**.  
El peso inferido se considera objetivo secundario para **biomasa agregada por clase**.
```

### Variables disponibles en el dataset

El análisis utiliza **exclusivamente** las siguientes columnas, sin generar variables adicionales no documentadas:

| Variable | Columna |
|--------|--------|
| Experimento / día | `exp` |
| Peso real (g) | `peso_g` |
| Longitud real (mm) | `L_real` |
| Anchura real (mm) | `A_real` |
| Longitud en píxeles | `Longitud_px` |
| Anchura en píxeles | `Anchura_px` |
| Longitud estimada por visión (mm) | `L_av` |
| Anchura estimada por visión (mm) | `A_av` |
| Error relativo longitud (%) | `errL_pct` |
| Error relativo anchura (%) | `errA_pct` |
| Peso inferido (g) | `peso_pred` |
| Error absoluto peso (g) | `err_abs_peso` |
| Error relativo peso (%) | `err_rel_peso` |

## Carga y validación inicial del dataset

### Carga de datos

In [10]:
# =========================
# 0) Imports y configuración
# =========================
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as st

import statsmodels.api as sm
import statsmodels.stats.api as sms
from sklearn.metrics import r2_score

# --- Rutas de entrada/salida (adaptadas para Jupyter-Book) ---
DATA_PATH = "../data/Dataset_validacion_final.xlsx"

# =========================
# 1) Carga y preprocesado
# =========================
df = pd.read_excel(DATA_PATH)

# Renombrado
rename_map = {
    "Experimento": "exp",
    "Longitud_av": "L_av",
    "Anchura_av": "A_av",
    "err_rel_Longitud": "errL_pct",
    "err_rel_Anchura": "errA_pct",
}
df = df.rename(columns=rename_map)

required = ["exp","peso_g","L_real","A_real","Longitud_px","Anchura_px","L_av","A_av",
            "errL_pct","errA_pct","peso_pred","err_abs_peso","err_rel_peso"]
missing = [c for c in required if c not in df.columns]
if missing:
    raise ValueError(f"Faltan columnas requeridas: {missing}")
df = df[required].copy()
df.head()


exp  peso_g  L_real  A_real  Longitud_px  Anchura_px   L_av   A_av  \
0    1    0.46      33      13          302         118  33.30  13.01   
1    5    0.46      33      13          299         118  32.96  13.01   
2    3    0.46      33      13          299         118  32.96  13.01   
3    3    0.67      39      15          351         136  38.70  14.99   
4    1    0.82      41      17          375         154  41.34  16.98   

   errL_pct  errA_pct  peso_pred  err_abs_peso  err_rel_peso  
0  0.909091  0.076923   0.483591      0.023591      5.128430  
1 -0.121212  0.076923   0.473780      0.013780      2.995702  
2 -0.121212  0.076923   0.473780      0.013780      2.995702  
3 -0.769231 -0.066667   0.748872      0.078872     11.771917  
4  0.829268 -0.117647   0.963993      0.143993     17.560066

### Verificación de la escala de errores relativos

In [11]:
# ----------------------------
# 2) Validación de escala de errores relativos
# ----------------------------
# Verificación: err_% debe coincidir con (estimado-real)/real*100
pct_L = (df["L_av"] - df["L_real"]) / df["L_real"] * 100.0
pct_A = (df["A_av"] - df["A_real"]) / df["A_real"] * 100.0
pct_P = (df["peso_pred"] - df["peso_g"]) / df["peso_g"] * 100.0

scale_notes = []
for col, pct in [("errL_pct", pct_L), ("errA_pct", pct_A), ("err_rel_peso", pct_P)]:
    x = df[col].astype(float).to_numpy()
    p = pct.astype(float).to_numpy()
    mask = np.isfinite(x) & np.isfinite(p)
    med_abs_direct = np.median(np.abs(x[mask] - p[mask]))
    med_abs_x100 = np.median(np.abs(x[mask]*100.0 - p[mask]))
    if med_abs_direct <= med_abs_x100:
        scale_notes.append(f"{col}: consistente con porcentaje (mediana |x - pct|={med_abs_direct:.3g}).")
    else:
        df[col] = df[col] * 100.0
        scale_notes.append(f"{col}: parecía fracción; convertido a porcentaje (×100).")

print("Verificación de escala (errores relativos):")
for s in scale_notes:
    print(" -", s)

Verificación de escala (errores relativos):
 - errL_pct: consistente con porcentaje (mediana |x - pct|=0).
 - errA_pct: consistente con porcentaje (mediana |x - pct|=0).
 - err_rel_peso: consistente con porcentaje (mediana |x - pct|=3.55e-15).


## Control de calidad y tratamiento de *outliers*

### Detección robusta de valores atípicos

Dado el carácter metrológico del estudio, se emplea un criterio robusto basado en z-score mediante MAD (Median Absolute Deviation) sobre:
 - error absoluto de longitud,
 - error absoluto de anchura,
 - error absoluto de peso.

Se marca como outlier cualquier observación con $∣𝑧_{MAD}|>5$

In [12]:
def mad_z(x):
    med = np.median(x)
    mad = np.median(np.abs(x - med))
    return 0.6745 * (x - med) / mad if mad > 0 else np.zeros_like(x)

df["err_abs_L"] = df["L_av"] - df["L_real"]
df["err_abs_A"] = df["A_av"] - df["A_real"]
df["err_abs_P"] = df["peso_pred"] - df["peso_g"]

df["z_L"] = mad_z(df["err_abs_L"])
df["z_A"] = mad_z(df["err_abs_A"])
df["z_P"] = mad_z(df["err_abs_P"])

df["outlier"] = (abs(df["z_L"]) > 5) | (abs(df["z_A"]) > 5) | (abs(df["z_P"]) > 5)

df_clean = df[~df["outlier"]].copy()
df.shape, df_clean.shape

((1250, 20), (1235, 20))

Se eliminan 15 observaciones (≈1.2%), quedando un conjunto final de 1235 registros, 
adecuado para análisis estadístico robusto sin distorsionar el comportamiento global.

In [5]:
# =========================
# 1) Carga y preprocesado
# =========================
df = pd.read_excel(DATA_PATH)

# Validación de columnas
missing = [c for c in COLS.values() if c not in df.columns]
if missing:
    raise ValueError(f"Faltan columnas requeridas: {missing}")

# Forzar tipos numéricos
for c in COLS.values():
    df[c] = pd.to_numeric(df[c], errors="coerce")

# Eliminar filas con datos incompletos
n0 = len(df)
df = df.dropna(subset=list(COLS.values())).copy()
n1 = len(df)

# Residuales (visión - real) y errores absolutos
df["errL_mm"] = df[COLS["L_av"]] - df[COLS["L_real"]]
df["errA_mm"] = df[COLS["A_av"]] - df[COLS["A_real"]]
df["errW_g"]  = df[COLS["peso_pred"]] - df[COLS["peso_real"]]

df["abs_errL_mm"] = df["errL_mm"].abs()
df["abs_errA_mm"] = df["errA_mm"].abs()
df["abs_errW_g"]  = df["errW_g"].abs()

print(f"Filas iniciales: {n0} | Filas tras limpiar nulos: {n1}")
print("Experimentos detectados:", sorted(df[COLS["exp"]].unique()))
df.head(3)


Filas iniciales: 1250 | Filas tras limpiar nulos: 1250
Experimentos detectados: [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5)]


peso_g  L_real  A_real  Longitud_px  Anchura_px   L_av   A_av  errL_pct  \
0    0.46      33      13          302         118  33.30  13.01  0.909091   
1    0.46      33      13          299         118  32.96  13.01 -0.121212   
2    0.46      33      13          299         118  32.96  13.01 -0.121212   

   errA_pct  exp  peso_pred  err_abs_peso  err_rel_peso  errL_mm  errA_mm  \
0  0.076923    1   0.483591      0.023591      5.128430     0.30     0.01   
1  0.076923    5   0.473780      0.013780      2.995702    -0.04     0.01   
2  0.076923    3   0.473780      0.013780      2.995702    -0.04     0.01   

     errW_g  abs_errL_mm  abs_errA_mm  abs_errW_g  
0  0.023591         0.30         0.01    0.023591  
1  0.013780         0.04         0.01    0.013780  
2  0.013780         0.04         0.01    0.013780

El conjunto de datos incluye, para cada individuo, errores relativos asociados a las estimaciones de longitud, anchura y peso (errL_pct, errA_pct, err_rel_peso), así como errores absolutos expresados en unidades físicas (abs_errL_mm, abs_errA_mm, abs_errW_g). Los errores absolutos se interpretan directamente como la diferencia entre la medida estimada por visión artificial y la medida de referencia, y constituyen la base principal para la evaluación metrológica del sistema y para el análisis del impacto de la incertidumbre sobre la clasificación por talla.

Los errores relativos de longitud y anchura se encuentran ya expresados en porcentaje (%) en el dataset original y pueden presentar valores positivos o negativos, reflejando respectivamente situaciones de sobreestimación o subestimación de la medida real. Dado que los errores absolutos observados son muy reducidos en términos de magnitud física, los valores relativos pueden ser numéricamente pequeños (por ejemplo, del orden de décimas de porcentaje), lo cual es coherente con un alto nivel de precisión del sistema de visión artificial. En consecuencia, estos errores relativos se utilizaron tal y como fueron proporcionados, sin aplicar ningún proceso adicional de normalización o reescalado, preservando así tanto su magnitud como su signo.

Con el fin de evitar interpretaciones ambiguas y garantizar la coherencia del análisis estadístico, se verificó previamente que todas las columnas de error relativo estuvieran expresadas de forma consistente en porcentaje y dentro de rangos razonables, descartándose la presencia de valores expresados como fracción. Adicionalmente, para aquellos análisis centrados exclusivamente en la magnitud del error y no en su dirección, se consideró el uso del valor absoluto del error relativo como variable complementaria, manteniendo siempre la versión firmada para el estudio de posibles sesgos sistemáticos.

### Normalización de errores relativos

El conjunto de datos incluye errores relativos asociados a las estimaciones de longitud, anchura y peso (`errL_pct`, `errA_pct`, `err_rel_peso`). En función del origen y del proceso de registro, estos errores pueden encontrarse expresados indistintamente como fracción o como porcentaje. Con el fin de evitar ambigüedades en la interpretación y garantizar la coherencia metrológica del análisis estadístico, se implementó un procedimiento automático de detección de escala, normalizando todos los errores relativos a porcentaje (%) cuando fuese necesario.

In [6]:
def normalize_pct(series: pd.Series):
    """Normaliza fracción -> porcentaje si mediana(|x|) < 0.5."""
    s = series.astype(float).copy()
    med = np.nanmedian(np.abs(s.values))
    if med < 0.5:
        return s * 100.0, True
    return s, False

df["errL_pct_norm"], convL = normalize_pct(df[COLS["errL_pct"]])
df["errA_pct_norm"], convA = normalize_pct(df[COLS["errA_pct"]])
df["errW_pct_norm"], convW = normalize_pct(df[COLS["err_rel_peso"]])

print("Normalización aplicada:", {"errL_pct": convL, "errA_pct": convA, "err_rel_peso": convW})
df[[COLS["errL_pct"], "errL_pct_norm", COLS["errA_pct"], "errA_pct_norm", COLS["err_rel_peso"], "errW_pct_norm"]].head(3)

Normalización aplicada: {'errL_pct': True, 'errA_pct': True, 'err_rel_peso': False}


errL_pct  errL_pct_norm  errA_pct  errA_pct_norm  err_rel_peso  \
0  0.909091      90.909091  0.076923       7.692308      5.128430   
1 -0.121212     -12.121212  0.076923       7.692308      2.995702   
2 -0.121212     -12.121212  0.076923       7.692308      2.995702   

   errW_pct_norm  
0       5.128430  
1       2.995702  
2       2.995702

## Variables disponibles

El dataset contiene:

- `Peso (g)` : peso real (variable objetivo para validación)
- `Longitud_REAL (mm)`, `Anchura_REAL (mm)` : medidas reales (referencia)
- `Longitud_AV (mm)`, `Anchura_AV (mm)` : medidas obtenidas por visión artificial (AV)
- `Error_longitud_% (AV-REAL)`, `Error_anchura_% (AV-REAL)` : errores relativos metrológicos ya calculados
- `Longitud_px`, `Anchura_px` : medidas en píxeles (no utilizadas aquí para el modelo; el modelo opera en mm)
- `Experimento` : identificador construido (E1..E5)

### Enfoque
1. Ajuste del modelo con medidas **REAL** (para estimación robusta de parámetros).
2. Inferencia operativa usando dimensiones **AV**.
3. Evaluación del error relativo del peso inferido frente al peso real.


In [ ]:
# --- Preparación: comprobaciones y columnas logarítmicas ---
COL_W = "Peso (g)"
L_real = "Longitud_REAL (mm)"
A_real = "Anchura_REAL (mm)"
L_av = "Longitud_AV (mm)"
A_av = "Anchura_AV (mm)"

# Verificar que todas las filas son válidas para log (positivas y no nulas)
mask = (
    df[[COL_W, L_real, A_real, L_av, A_av]].notna().all(axis=1) &
    (df[[COL_W, L_real, A_real, L_av, A_av]] > 0).all(axis=1)
)
assert mask.all(), "Hay filas no válidas para log; revisa el dataset."

# Logs (para modelo alométrico)
df["_log_w"]   = np.log(df[COL_W])
df["_log_Lr"]  = np.log(df[L_real])
df["_log_Ar"]  = np.log(df[A_real])
df["_log_Lav"] = np.log(df[L_av])
df["_log_Aav"] = np.log(df[A_av])

df[[COL_W, L_real, A_real, L_av, A_av, "Experimento"]].head()


## Modelo alométrico (log–log) y corrección de retransformación

Se ajusta un modelo lineal en el espacio logarítmico:

\[
\log(P) = b_0 + b_1 \log(L) + b_2 \log(A) + \varepsilon
\]

donde:

- \(P\) es el peso (g),
- \(L\) longitud (mm),
- \(A\) anchura (mm).

La predicción en espacio real se obtiene como:

\[
\hat{P} = \exp(\hat{y})
\]

y, para reducir el sesgo por retransformación, se aplica una corrección multiplicativa tipo **smearing**:

\[
\hat{P}_{corr} = \exp(\hat{y}) \cdot S, \quad S = \frac{1}{n}\sum_i \exp(\hat{\varepsilon}_i)
\]

Este enfoque es especialmente útil en producción porque:
- ofrece **interpretabilidad** y **estabilidad**,
- es **eficiente computacionalmente** (Edge),
- facilita auditoría y recalibraciones controladas si fuese necesario.


In [ ]:
# --- Ajuste del modelo con dimensiones REAL (global) ---
Xr = sm.add_constant(df[["_log_Lr", "_log_Ar"]])
y = df["_log_w"]

model = sm.OLS(y, Xr).fit()
print(model.summary())

b0 = float(model.params["const"])
b1 = float(model.params["_log_Lr"])
b2 = float(model.params["_log_Ar"])

smearing = float(np.mean(np.exp(model.resid)))

k = float(np.exp(b0))
k_corr = k * smearing

print("\nParámetros:")
print(f"b0={b0:.6f}, b1={b1:.6f}, b2={b2:.6f}")
print(f"Smearing S={smearing:.6f}")
print("\nFórmula operativa (smearing incluido):")
print(f"Peso(g) = {k_corr:.12f} * L^{b1:.6f} * A^{b2:.6f}   (L,A en mm)")


In [ ]:
# --- Inferencia de peso usando dimensiones REAL y AV ---
def pred_weight_from_logs(logL, logA):
    return np.exp(b0 + b1 * logL + b2 * logA)

# Predicción (naive) y corregida (smearing)
df["Peso_inferido_REAL"] = pred_weight_from_logs(df["_log_Lr"], df["_log_Ar"])
df["Peso_inferido_REAL_smear"] = df["Peso_inferido_REAL"] * smearing

df["Peso_inferido_AV"] = pred_weight_from_logs(df["_log_Lav"], df["_log_Aav"])
df["Peso_inferido_AV_smear"] = df["Peso_inferido_AV"] * smearing

# Error relativo (%)
df["Error_rel_peso_REAL_pct"] = (np.abs(df["Peso_inferido_REAL_smear"] - df[COL_W]) / df[COL_W]) * 100
df["Error_rel_peso_AV_pct"]   = (np.abs(df["Peso_inferido_AV_smear"] - df[COL_W]) / df[COL_W]) * 100

df[[COL_W, "Peso_inferido_AV_smear", "Error_rel_peso_AV_pct"]].head()


## Tablas de resultados (estadísticos y coberturas)

Se generan tablas para:
- estadísticos de peso real,
- error metrológico de longitud/anchura (AV-REAL),
- error del peso inferido con AV,
- coberturas por umbrales (≤5%, ≤10%, ...),
- resultados por experimento (repetibilidad inter-día),
- resultados por rangos de peso (robustez por tamaño).


In [ ]:
def desc(s: pd.Series) -> pd.Series:
    return pd.Series({
        "N": int(s.count()),
        "Media": float(s.mean()),
        "Mediana": float(s.median()),
        "Desv.Típ.": float(s.std(ddof=1)),
        "P5": float(s.quantile(0.05)),
        "P95": float(s.quantile(0.95)),
        "Mín": float(s.min()),
        "Máx": float(s.max()),
    })

tab_peso = desc(df[COL_W]).to_frame("Peso (g)")
tab_errL = desc(df["Error_longitud_% (AV-REAL)"]).to_frame("Error_longitud_% (AV-REAL)")
tab_errA = desc(df["Error_anchura_% (AV-REAL)"]).to_frame("Error_anchura_% (AV-REAL)")
tab_errW_AV = desc(df["Error_rel_peso_AV_pct"]).to_frame("Error_rel_peso_AV_pct")

tab_peso.round(3), tab_errW_AV.round(3)


In [ ]:
# Cobertura acumulada por umbrales de error relativo (AV)
thresholds = [5, 10, 15, 20]
tab_cov = pd.DataFrame([{
    "Umbral (%)": t,
    "% individuos (<=umbral)": float((df["Error_rel_peso_AV_pct"] <= t).mean() * 100),
    "N (<=umbral)": int((df["Error_rel_peso_AV_pct"] <= t).sum()),
} for t in thresholds])

tab_cov.round(2)


In [ ]:
# Resumen por experimento (día)
per = df.groupby("Experimento")

tab_per_metrology = per.agg({
    "Error_longitud_% (AV-REAL)": ["mean", "median", lambda s: s.quantile(0.95)],
    "Error_anchura_% (AV-REAL)": ["mean", "median", lambda s: s.quantile(0.95)],
})
tab_per_metrology.columns = ["L_mean", "L_med", "L_P95", "A_mean", "A_med", "A_P95"]

tab_per_weight = per["Error_rel_peso_AV_pct"].agg(["count", "mean", "median", lambda s: s.quantile(0.95), "max"])
tab_per_weight.columns = ["N", "Media_%", "Mediana_%", "P95_%", "Max_%"]

tab_per_metrology.round(4), tab_per_weight.round(2)


In [ ]:
# Error por rangos de peso real (robustez por tamaño)
bins = [-np.inf, 2, 5, 10, np.inf]
labels = ["<2 g", "2–5 g", "5–10 g", ">=10 g"]

df["Rango_peso"] = pd.cut(df[COL_W], bins=bins, labels=labels)

grp = df.groupby("Rango_peso", observed=True)["Error_rel_peso_AV_pct"]
tab_bins = pd.DataFrame({
    "N": grp.count().astype(int),
    "Media_%": grp.mean(),
    "Mediana_%": grp.median(),
    "P95_%": grp.quantile(0.95),
    "Max_%": grp.max(),
}).round(2)

tab_bins


## Figuras (generación reproducible)

Se exportan las figuras a una carpeta local (`jbook_assets_exp`) para poder referenciarlas desde Jupyter-Book.


In [ ]:
# --- Generación de figuras ---
out_dir = "jbook_assets_exp"
os.makedirs(out_dir, exist_ok=True)

# Figura 4-1: distribución errores dimensionales
plt.figure()
plt.hist(df["Error_longitud_% (AV-REAL)"], bins=40, alpha=0.7, label="Error longitud (%)")
plt.hist(df["Error_anchura_% (AV-REAL)"], bins=40, alpha=0.7, label="Error anchura (%)")
plt.xlabel("Error relativo (%)")
plt.ylabel("Frecuencia")
plt.title("Distribución del error relativo en medidas AV vs REAL (5 experimentos)")
plt.legend()
fig1 = os.path.join(out_dir, "fig_4_1_error_dimensiones_5exp.png")
plt.tight_layout(); plt.savefig(fig1, dpi=200); plt.close()

# Figura 4-2: histograma error de peso AV
plt.figure()
plt.hist(df["Error_rel_peso_AV_pct"], bins=50)
plt.xlabel("Error relativo del peso inferido (%)")
plt.ylabel("Frecuencia")
plt.title("Histograma del error relativo del peso inferido (dimensiones AV)")
fig2 = os.path.join(out_dir, "fig_4_2_hist_error_peso_5exp.png")
plt.tight_layout(); plt.savefig(fig2, dpi=200); plt.close()

# Figura 4-3: boxplot error por experimento
plt.figure()
data_box = [df.loc[df["Experimento"] == i, "Error_rel_peso_AV_pct"].values for i in range(1, 6)]
plt.boxplot(data_box, labels=[f"E{i}" for i in range(1, 6)], showfliers=True)
plt.xlabel("Experimento (día)")
plt.ylabel("Error relativo del peso inferido (%)")
plt.title("Variabilidad inter-experimento del error de peso (AV)")
fig3 = os.path.join(out_dir, "fig_4_3_boxplot_error_peso_por_experimento.png")
plt.tight_layout(); plt.savefig(fig3, dpi=200); plt.close()

# Figura 4-4: boxplot por rango de peso
plt.figure()
data_box = [df.loc[df["Rango_peso"] == lbl, "Error_rel_peso_AV_pct"].values for lbl in ["<2 g", "2–5 g", "5–10 g", ">=10 g"]]
plt.boxplot(data_box, labels=["<2 g", "2–5 g", "5–10 g", ">=10 g"], showfliers=True)
plt.xlabel("Rango de peso real")
plt.ylabel("Error relativo del peso inferido (%)")
plt.title("Error de peso inferido (AV) por rangos de peso real")
fig4 = os.path.join(out_dir, "fig_4_4_boxplot_error_por_rango_peso.png")
plt.tight_layout(); plt.savefig(fig4, dpi=200); plt.close()

# Figura 4-5: scatter peso inferido vs real por experimento
plt.figure()
for i in range(1, 6):
    sub = df[df["Experimento"] == i]
    plt.scatter(sub[COL_W], sub["Peso_inferido_AV_smear"], s=10, label=f"E{i}", alpha=0.7)
mx = float(np.nanmax([df[COL_W].max(), df["Peso_inferido_AV_smear"].max()]))
plt.plot([0, mx], [0, mx])
plt.xlabel("Peso real (g)")
plt.ylabel("Peso inferido (g)")
plt.title("Peso inferido vs real por experimento (AV)")
plt.legend()
fig5 = os.path.join(out_dir, "fig_4_5_scatter_peso_real_vs_inferido_por_experimento.png")
plt.tight_layout(); plt.savefig(fig5, dpi=200); plt.close()

# Figura 4-6: evolución media y P95 por experimento
plt.figure()
x = np.arange(1, 6)
plt.plot(x, tab_per_weight["Media_%"].values, marker="o", label="Media (%)")
plt.plot(x, tab_per_weight["P95_%"].values, marker="o", label="P95 (%)")
plt.xlabel("Experimento (día)")
plt.ylabel("Error relativo (%)")
plt.title("Evolución de error de peso (media y P95) por experimento")
plt.legend()
fig6 = os.path.join(out_dir, "fig_4_6_media_p95_por_experimento.png")
plt.tight_layout(); plt.savefig(fig6, dpi=200); plt.close()

[fig1, fig2, fig3, fig4, fig5, fig6]


## Discusión técnico-operativa (orientación CDTI)

### Lecturas principales (resumen)

- **Metrología (AV vs REAL):** los errores relativos de longitud y anchura permanecen contenidos y consistentes a lo largo de 5 días, lo que aporta evidencia de **repetibilidad del pipeline de visión** con configuración constante.
- **Error de peso inferido (AV):** el error relativo presenta estabilidad central compatible con clasificación por tamaños. La dispersión en el extremo bajo (<2 g) es esperable por naturaleza geométrica y biológica.
- **Repetibilidad inter-día:** la comparación por experimento (media/mediana/P95) permite argumentar **robustez temporal sin recalibración**, aspecto especialmente relevante para justificar TRL en memoria CDTI.
- **Criterios de aceptación recomendados para clasificación:** mediana ≤10%, cobertura ≤10% ≥55–65%, P95 ≤25%, apoyado por QA para casos atípicos.

### TRL (síntesis)
Con la evidencia multi-día y el modelo único sin recalibración, la tecnología se sitúa en **TRL 6–7**, dependiendo del grado de equivalencia del banco de ensayo con la línea real (cadencia, manipulación, iluminación, etc.).

### Recomendación para reforzar TRL 7→8
- Añadir trazas operativas del entorno (parámetros de captura, control QA, incidencias)
- Pilotos adicionales bajo variabilidad controlada (lote, densidad, iluminación)
- Integración final con criterio de clasificación (umbrales por categoría) y medición de tasa de recaptura/“errores”


In [ ]:
# (Opcional) Exportar un Excel enriquecido con columnas calculadas para auditoría/QA
out_xlsx = "Datos_experimentos_lenguado_enriquecido.xlsx"
df.to_excel(out_xlsx, index=False)
out_xlsx
